### DISCUSSION TITLE
Ability to add "delta" changes only to an existing sqitch script/namespace

### DISCUSSION THREAD

**Discussion by 1x-eng** [theory/sqitch#708] [Q&A] [answered]
Hello, 

Does sqitch support staging "delta" changes to existing scripts/namespaces without needing to do rebase? Or another alternative is to somehow register the new script with an existing one so they become part of single atomic transaction?

A vanilla example:

1. `sqitch add tables` creates new "namespace" of tables and tables.sql in deploy, revert and verify. Lets say this was deployed successfully 
2. Later, we have a requirement which makes sense to be part of the same `tables.sql` script to avoid fragmentation + also make sure statements that should be part of single transaction live together in code. Adding new SQL statements to existing `tables.sql` does not get picked by `sqitch deploy` as it goes about suggesting `Nothing to deploy`. Same with `sqitch status`, it does not detect the newly added change (im looking for a `git diff` kinda functionality if it helps simplify my ask). 

So my ask is:
1. Is there a way to detect this delta change? and somehow deploy only a specific namespace (looking for something like `terraform --target` if that helps?)
2. If not, is there a way to associate a new sqitch script to an existing one? (is `--requires` the best bet? Im presuming `requires` does not guarantee that everything in new script will be part of same atomic transaction?

Appreciate any help/ resources if this was previously discussed/answered. Thank you!

**Comment by theory** [comment, marked as answer]
No, it doesn't work quite like that. One of the ideas behind Sqitch is that there are development cycles and releases. During development, the expectation is that you can modify and reorganize your changes however you like. If you've made a `tables` change and want to, say, add a column or index, you can revert the change, made the changes, then deploy it again. Ideally you're using a disposable database for this. Likewise for any CI pipelines.

Then, once it's reached a stable point, you [`sqitch tag`](https://sqitch.org/docs/manual/sqitch-tag/), bundle things up, and release it. The idea is that you _never_ modify any changes prior to that tag — because it has been released, and as you point out, `sqitch deploy` won't notice any changes to them.

Now say it's after a tagged release, and you realize you need to add a column to a table. The idea here is you create a new change, `add_column` or some such. If you can make an idempotent change, such as `CREATE OR REPLACE VIEW`, you can use [`sqitch rework`](https://sqitch.org/docs/manual/sqitch-rework/) to create a copy of a previous change and then edit it inline. Again, make all the changes you want, revert and redeploy (or, better, [`rebase`](https://sqitch.org/docs/manual/sqitch-rebase/) until you're ready to make the next release.

So, to your questions:

1. No, there is no way to detect this delta change. That might change in the future; this is the idea behind `sqitch merge`, as described in #200, and we have a SHA hash of the deploy script stored in the database, so modifications can be detected. But it hasn't been a priority to develop it.

2. Yes, `--requires` is the way to go, especially a non-idempotent change. It's slightly redundant, since changes in a single plan are always applied in order, but useful documentation. (Requires becomes more powerful for cross-project dependencies.) The other way is to use `sqitch rework` for idempotent changes. I've known people to use them even for non-idempotent changes like adding a column, though that honestly always felt a little weird to me.

Sqitch does not enforce transactions in change scripts, but as each script is sent to the database individually via the database-native CLI (`psql`, `mysql`, `sqlplus`, etc.), you can control the transactions _within that script_. I generally try to make each script a single transaction, but it's not always possible, especially for databases that don't support transactional DDL.

HTH!
